src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/Frame.cs

```diff
@@ -552,9 +572,22 @@ public void Flush()
             Output.Flush();
         }
 
-        public async Task FlushAsync(CancellationToken cancellationToken)
+        public Task FlushAsync(CancellationToken cancellationToken)
         {
-            await InitializeResponse(0);
+            var initializeTask = InitializeResponseMayAwait(0);
+            // If return is TaskCache.CompletedTask no awaiting is required
+            if (!ReferenceEquals(initializeTask, TaskCache.CompletedTask))
+            {
+                return FlushAsyncAwaited(initializeTask, cancellationToken);
+            }
+
+            return Output.FlushAsync(cancellationToken);
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
```

Comment:
Do you need the attribute? An async method has a ton of IL. Won't that mean there is already zero chance of it being inlined?